docs/plugins.md

```diff
@@ -0,0 +1,201 @@
+
+`Scope` is a logical group of files; for instance `source` files, or each component's individual scope (XML definition, related components and imported scripts). A file can be included in multiple scopes.
+
+`BrsFile` is a `.brs` or `.bs` file; its AST can be modified by a plugin.
+
+`XmlFile` is a component; it doesn't have a proper AST currently but it can be "patched".
+
+```typescript
+// ProgramBuilder events:
+interface ProgramBuilderEventMap {
+    'program-created': (program: Program) => void;
+    'before-prepublish': (files: SourceDestEntry[]) => void;
+    'before-publish': (files: SourceDestEntry[]) => void;
+    'after-publish': (files: SourceDestEntry[]) => void;
+}
+
+// Program events:
+interface ProgramEventMap {
+    'validate-start': () => void;
+    'validate-end': () => void;
+    'scope-created': (scope: Scope) => void;
+    'scope-disposed': (scope: Scope) => void;
+    'file-source-loaded': (source: SourceInfo) => void;
+    'file-parsed': (file: BrsFile) => void;
+    'file-validated': (file: BrsFile) => void;
+    'component-source-loaded': (source: SourceInfo) => void;
+    'component-parsed': (file: XmlFile) => void;
+    'component-validated': (file: XmlFile) => void;
+    'before-transpile': (entries: TranspileEntry[]) => void;
+    'after-transpile': (entries: TranspileEntry[]) => void;
+}
+
+// Scope events:
+interface ScopeEventsMap {
+    'invalidated': () => void;
+    'validate-start': ValidateHandler;
+    'validate-end': ValidateHandler;
+}
+
+// related types:
+interface SourceDestEntry {
+    src: string;
+    dest: string;
+}
+
+interface SourceInfo {
+    pathAbsolute: string;
+    src: string;
+}
+
+interface TranspileEntry {
+    file: (BrsFile | XmlFile);
+    outputPath: string;
+}
+
+type ValidateHandler = (scope: Scope, files: (BrsFile | XmlFile)[], callables: CallableContainerMap) => void;
+interface CallableContainerMap {
+    [name: string]: CallableContainer[];
+}
+interface CallableContainer {
+    callable: Callable;
+    scope: Scope;
+}
+```
+
+## Plugin API
+
+Plugins are JavaScript modules dynamically loaded by the compiler. Their entry point is a required `initPlugin` function.
+
+To walk/modify the AST, a number of helpers are provided in `brighterscript/dist/parser/ASTUtils`.
+
+### Working with TypeScript
+
+It is highly recommended to use TypeScript as intellisence helps greatly writing  code against new APIs.
+
+```bash
+# install modules needed to compile a plugin
+npm install brighterscript typescript @types/node
+
+# transpile to JS (with source maps for debugging)
+npx tsc myPlugin.ts --sourceMap
+
+# add --watch for continuous transpilation
+npx tsc myPlugin.ts --sourceMap --watch
+```
+
+### Example diagnostic plugins
+
+```typescript
+// myDiagnosticPlugin.ts
+import { ProgramBuilder, Scope, BrsFile,  XmlFile, CallableContainerMap } from 'brighterscript';
+
+// entry point
+export function initPlugin(builder: ProgramBuilder) {
+    builder.on('program-created', program => {
```

Comment:
Well that was basically the initial API :D

I think it's harder to use/discover for plugin authors, and to scale to more events and objects, and I didnt want to introduce API changes otherwise.

I also don't like the way the initial source creation doesn't trigger an event though, but I couldn't think of a clean solution with the events as it happens in the constructor. Maybe it could be delayed to the first source file registration.

Reply:
What about "bubbling" the events maybe? E.g. what if you could listen to all the events on the builder instance? And I'll figure something clean enough to have the source created event.

Reply:
Bubbling the events would work, and does simplify things a bit. However, why is my suggestion of a default export harder for plugin authors to scale? If we exported an interface like this:
```typescript
export interface Plugin {
    'program-created'?: (program: Program) => void;
    'scope-created'?: (scope: Scope) => void;
    'scope-validate-start'?: (scope: Scope) => void;
}
```

Then plugin authors could write their plugins like this: 
```typescript
// myDiagnosticPlugin.ts
export default {
    'program-created': program => { },
    'scope-created': scope => { },
    'scope-validate-start': scope => { }
} as Plugin;
```
The type casting at the end would give the plugin developers full type checking and discoverability for all plugin features. We could use the same `Plugin` interface internally to track all of the plugins, so as we add new plugin features, we would simply update the interface and both our internal code and the plugin api get the changes together. 

Honestly, I think this approach provides _easier_ discoverability for plugin features than events scattered across the various BrighterScript internals (but bubbling the events would solve that).

Another benefit to my proposed approach is the plugins become easier to test / mock, since they're just plain functions. Not a big deal, but definitely something to consider.